### PR TITLE
[Java] Add java api overload doc and test

### LINKED ISSUE #14186: [Java] Ray call java api doesn't support method overload
### What is the problem?
Ray java api use method reference(lambda) to reference a method which will be executed remotely and enforce type safety in compile-time. But when the referenced method is overloaded, the compiler will complain. This is because both ray call api(`io.ray.RayCall#task/io.ray.ActorCall#task`) and referenced methods are overloaded, neither side of the methods are determinate ahead, the compiler can't determinate the methods of any side. 
For static function call, this can be mitigated if the users add more type informations in code like:
image

For actor task call, this can be mitigated too:
image

And neither the static method call or instance methods call can work when static methods are overlaoded with instance methods and the first parameter type is the same with actor class type which is uncommon:
image

All those limits are exposed by the methods reference. The ray runtime does support method overload, but by the limitation of java language we can't reference overloaded method:
image
`Reference to an instance method of an arbitrary object of a particular type` doesn't support `overload`.

### Reproduction (REQUIRED)
image
image

### Solution
* For both normal task call(static function call) and actor task call(instance function call), method overload can be mitigated if the users add more type informations. When static methods are overlaoded with instance methods and the first parameter type is the same with actor class type which is uncommon, this solution may not work. But since it's uncommon, it's OK we dones't support this case.

* Both those call can be resolved by adding a non-compile-time type-safety task call interface for overloaded methods and put method resolution and linking into runtime when the function starts to execute. The method resolution will be based on the arguments number and types of passed. API example: `<R> ActorTaskCaller<R> task(String methodName, Object... args)`. The cons are we lose type-safety and have some performance cost because the method  resolution are put into runtime instead of compile-time.

### Reference
1. https://dzone.com/articles/java-8s-method-references-put-further-restrictions
2. https://docs.oracle.com/javase/tutorial/java/javaOO/methodreferences.html
- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by chaokunyang**
## Why are these changes needed?
Add java api overloaded methods call doc and tests.

## Related issue number
closes #14186

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by chaokunyang** [comment]
image

File: doc/source/advanced.rst [hunk 1]
```diff
@@ -206,6 +206,68 @@ Ray supports resource specific accelerator types. The `accelerator_type` field c
 See `ray.util.accelerators` to see available accelerator types. Current automatically detected accelerator types include Nvidia GPUs.
 
 
+Overloaded Functions
+--------------------
+Ray java api support call overloaded java functions remotely. For example, consider the following.
```
Comment (raulchen) [inline doc/source/advanced.rst:211]: worth mentioning the reason here. and some minor grammar issues.
```
Ray Java API supports calling overloaded java functions remotely. However, due to the limitation of Java compiler type inference, one must explicitly cast the method reference to the correct function type. ...
```

Comment (raulchen) [reply doc/source/advanced.rst:211]: Just found that you've mentioned this below. I think better put it here.

**Review by raulchen — APPROVED** [review]

File: doc/source/advanced.rst (hunk 1 above)
Comment (chaokunyang) [reply doc/source/advanced.rst]: fixed

**Pull request merged**